fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,476 @@
+
+type ParentViewPayload = (BlockHash, ValidatorSet, Vec<CrossMsg>);
+
+/// The default parent finality provider
+pub struct InMemoryFinalityProvider {
+    config: Config,
+    parent_view_data: ParentViewData,
+    /// This is a in memory view of the committed parent finality,
+    /// it should be synced with the store committed finality, owner of the struct should enforce
+    /// this.
+    last_committed_finality: TVar<IPCParentFinality>,
+}
+
+/// Tracks the data from the parent
+#[derive(Clone)]
+struct ParentViewData {
+    height_data: TVar<SequentialKeyCache<BlockHeight, ParentViewPayload>>,
+}
+
+impl ParentViewData {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>, Error> {
+        let cache = self.height_data.read()?;
+        // safe to unwrap, we dont allow no upper bound
+        Ok(cache.upper_bound())
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<BlockHash>, Error> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.0.clone()))
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>, Error> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.1.clone()))
+    }
+
+    fn top_down_msgs(
+        &self,
+        from_height: BlockHeight,
+        to_height: BlockHeight,
+    ) -> StmResult<Vec<CrossMsg>, Error> {
+        let cache = self.height_data.read()?;
+        let v = cache
+            .values_within(from_height, to_height)
+            .flat_map(|i| i.2.iter())
+            .cloned()
+            .collect();
+        Ok(v)
+    }
+}
+
+impl ParentViewProvider for InMemoryFinalityProvider {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>, Error> {
+        let h = self.parent_view_data.latest_height()?;
+        Ok(h)
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<BlockHash>, Error> {
+        let v = self.parent_view_data.block_hash(height)?;
+        Ok(v)
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>, Error> {
+        let v = self.parent_view_data.validator_set(height)?;
+        Ok(v)
+    }
+
+    fn top_down_msgs(&self, height: BlockHeight) -> StmResult<Vec<CrossMsg>, Error> {
+        let v = self.parent_view_data.top_down_msgs(height, height)?;
+        Ok(v)
+    }
+
+    fn new_parent_view(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+        top_down_msgs: Vec<CrossMsg>,
+    ) -> StmResult<(), Error> {
+        if !top_down_msgs.is_empty() {
+            // make sure incoming top down messages are ordered by nonce sequentially
+            ensure_sequential_by_nonce(&top_down_msgs)?;
+        };
+
+        let r = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache
+                .append(height, (block_hash, validator_set, top_down_msgs))
```

Comment:
Based on what I have seen so far of the cache, this will fail unless `height` follows the previous by `increment`, which we know is 1 in the syncer, but could be anything here.

Whose responsibility is it to aggregate into the expected intervals? This or the caller?

Reply:
Oh I see it is indeed hardcoded for single blocks [here](https://github.com/consensus-shipyard/fendermint/pull/206/files#diff-8edd9923a9eb262950fd919a858ece4449d8a41abf401bd3c1e4ac1d63190454R171).

Maybe worth a comment here to say this is the assumption in this method.

Reply:
comment added

Reply:
Can't see where but I believe you.